# 웹 크롤링1 - Static Crawling
201921529 김채연

# 1. urllib
- 파이썬은 웹 사이트에 있는 데이터를 추출하기 위해 'urllib' 라이브러리 사용
- 이를 이용해 HTTP 또는 FTP를 사용해 데이터 다운로드 가능
- urllib은 URL을 다루는 모듈을 모아 놓은 패키지
- urllib.request 모듈은 웹 사이트에 있는 데이터에 접근하는 기능 제공, 또한 인증, 리다렉트, 쿠키처럼 인터넷을 이용한 다양한 요청과 처리가 가능

In [3]:
# urllib 라이브러리에서 request 모듈 불러오기
from urllib import request

## 1.1. urllib.request를 이용한 다운로드
- urllib.request 모듈에 있는 urlretrieve() 함수 이용
- 다음의 코드는 PNG 파일을 test.png 라는 이름의 파일로 저장하는 예제임

In [4]:
# 라이브러리 읽어들이기 
from urllib import request

url="http://uta.pw/shodou/img/28/214.png"
savename="test.png"

# url에 저장된 png파일을 test.png라는 이름의 파일로 저장하기
request.urlretrieve(url, savename)
print("저장되었습니다")

저장되었습니다


## 1.2. urlopen으로 파일에 저장하는 방법
- request.urlopen()은 메모리에 데이터를 올린 후 파일에 저장하게 된다.

In [5]:
# URL과 저장경로 지정하기
url = "http://uta.pw/shodou/img/28/214.png"
savename = "test1.png"
#다운로드
mem = request.urlopen(url).read()
#파일로 저장하기, wb는 쓰기와 바이너리모드
with open(savename, mode="wb") as f:
    f.write(mem)
    print("저장되었습니다..")

저장되었습니다..


## 1.3. API 사용하기

### 클라이언트 접속 정보 출력 (기본)
- API는 사용자의 요청에 따라 정보를 반환하는 프로그램
- IP 주소, UserAgent 등 클라이언트 접속정보 출력하는 "IP 확인  API" 접근해서 정보를 추출하는 프로그램

In [6]:
#데이터 읽어들이기
url="http://api.aoikujira.com/ip/ini"
res=request.urlopen(url)
data=res.read()

#바이너리를 문자열로 변환하기
text=data.decode("utf-8")
print(text)

[ip]
API_URI=http://api.aoikujira.com/ip/get.php
REMOTE_ADDR=218.148.96.114
REMOTE_HOST=218.148.96.114
REMOTE_PORT=54958
HTTP_HOST=api.aoikujira.com
HTTP_USER_AGENT=Python-urllib/3.8
HTTP_ACCEPT_LANGUAGE=
HTTP_ACCEPT_CHARSET=
SERVER_PORT=80
FORMAT=ini




# 2. BeautifulSoup

- 스크레이핑(Scraping or Crawling)이란 웹 사이트에서 데이터를 추출하고, 원하는 정보를 추출하는 것을 의미
- **BeautifulSoup**란 파이썬으로 스크레이핑할 때 사용되는 라이브러리로서 HTML/XML에서 정보를 추출할 수 있도록 도와줌. 그러나 다운로드 기능은 없음.
- 파이썬 라이브러리는 pip 명령어를 이용해 설치 가능. Python Package Index(PyPI)에 있는 패키지 명령어를 한줄로 설치 가능
- URL (http://pypi.python.org/pypi)


In [11]:
pip install beautifulsoup4

Note: you may need to restart the kernel to use updated packages.


In [7]:
from bs4 import BeautifulSoup

In [8]:
# 예제 HTML

html = """
<html><body>
  <h1>스크레이핑이란?</h1>
  <p>웹 페이지를 분석하는 것</p>
  <p>원하는 부분을 추출하는 것</p>
</body></html>
"""

## 2.1. 기본 사용
- 다음은 Beautifulsoup를 이용하여 웹사이트로부터 HTML을 가져와 문자열로 만들어 이용하는 예제
- h1 태그를 접근하기 위해 html-body-h1 구조를 사용하여 soup.html.body.h1 이런식으로 이용하게 됨.
- p 태그는 두개가 있어 soup.html.body.p 한 후 next_sibling을 두번 이용하여 다음 p를 추출. 한번만 하면 그 다음 공백이 추출됨.
- HTML 태그가 복잡한 경우 이런 방식으로 계속 진행하기는 적합하지 않음.

### 2) HTML 분석하기

In [9]:
## BeautifulSoup으로 html소스를 python객체로 변환하기
## 첫 인자는 html소스코드, 두 번째 인자는 어떤 parser를 이용할지 명시.
## Python 내장 html.parser를 이용함.
soup = BeautifulSoup(html, 'html.parser')

### 3) 원하는 부분 추출하기

In [10]:
# html의 각 태그의 데이터에 접근하기
h1 = soup.html.body.h1
p1 = soup.html.body.p
p2 = p1.next_sibling.next_sibling

### 4) 요소의 글자 출력하기

In [11]:
# 위에서 제시된 예제 HTML의 각 요소의 글자 출력하기
print(f"h1 = {h1.string}")
print(f"p  = {p1.string}")
print(f"p  = {p2.string}")

h1 = 스크레이핑이란?
p  = 웹 페이지를 분석하는 것
p  = 원하는 부분을 추출하는 것


## 2.2. 요소를 찾는 method

### 단일 element 추출 : **find()**

BeautifulSoup는 루트부터 하나하나 요소를 찾는 방법 말고도 find()라는 메소드를 제공함

In [12]:
soup = BeautifulSoup(html, 'html.parser')

- 1) **find()** 메서드로 원하는 부분 추출하기

In [13]:
title = soup.find("h1")
body  = soup.find("p")

# h1 부분만 추출하기
print(title)

<h1>스크레이핑이란?</h1>


- 2) 텍스트 부분 출력하기

In [14]:
# h1의 텍스트 부분 출력하기
print(f"#title = {title.string}" )
# p의 텍스트 부분 출력하기
print(f"#body = {body.string}")

#title = 스크레이핑이란?
#body = 웹 페이지를 분석하는 것


### 복수 elements 추출: find_all()

여러개의 태그를 한번에 추출하고자 할때 사용함. 다음의 예제에서는 여러개의 태그를 추출하는 법을 보여주고 있음

In [15]:
# 예제 html
html = """
<html><body>
  <ul>
    <li><a href="http://www.naver.com">naver</a></li>
    <li><a href="http://www.daum.net">daum</a></li>
  </ul>
</body></html>
"""

soup = BeautifulSoup(html, 'html.parser')

- 1) **find_all()** 메서드로 추출하기

In [16]:
# "a" 해당 모든 태그를 불러오기
links = soup.find_all("a")
# 해당되는 태그들과 해당되는 태그들의 수 출력하기
print(links, len(links))

[<a href="http://www.naver.com">naver</a>, <a href="http://www.daum.net">daum</a>] 2


- 2) 링크 목록 출력하기

In [17]:
for a in links:
    href = a.attrs['href'] # attrs 속성에서 href의 속성에 있는 속성값을 추출
    text = a.string  # 문자열 추출하기
    print(text, ">", href)

naver > http://www.naver.com
daum > http://www.daum.net


# 3. Css Selector

Css Selector란, 웹상의 요소에 css를 적용하기 위한 문법으로, 즉 요소를 선택하기 위한 패턴입니다.

출처: https://www.w3schools.com/cssref/css_selectors.asp

복잡하게 구조화된 웹 사이트에서 자신이 원하는 데이터를 가져오기 위해서는 Css Selector에 대한 이해가 필요합니다.

|서식|설명|
|----|---|
|*|모든 요소를 선택|
|<요소 이름>|요소 이름을 기반으로 선택|
|<클래스 이름>|클래스 이름을 기반으로 선택|
| #<id 이름>|id 속성을 기반으로 선택|

##  BeautifulSoup에서 Css Selector 사용하기

BeautifulSoup에서는 Css Selector로 값을 가져올 수 있도록 find와는 다른 다음과 같은 메서드를 제공합니다.

|메서드|설명|
|------|----|
|soup.select_one(선택자)|CSS 선택자로 요소 하나를 추출합니다.|
|soup.select(선택자)|CSS 선택자로 요소 여러 개를 리스트를 추출합니다.|

In [18]:
html = """
<html><body>
<div id="meigen">
  <h1>위키북스 도서</h1>
  <ul class="items">
    <li>유니티 게임 이펙트 입문</li>
    <li>스위프트로 시작하는 아이폰 앱 개발 교과서</li>
    <li>모던 웹사이트 디자인의 정석</li>
  </ul>
</div>
</body></html>
"""

# HTML 분석하기 
soup = BeautifulSoup(html, 'html.parser')

- 필요한 부분을 CSS 쿼리로 추출하기

In [19]:
# 타이틀 부분 추출하기 --- (※3)
h1 = soup.select_one("div#meigen > h1").string
print(f"h1 = {h1}")

# 목록 부분 추출하기 --- (※4)
li_list = soup.select("div#meigen > ul.items > li")
for li in li_list:
  print(f"li = {li.string}")

h1 = 위키북스 도서
li = 유니티 게임 이펙트 입문
li = 스위프트로 시작하는 아이폰 앱 개발 교과서
li = 모던 웹사이트 디자인의 정석


# 4. 활용 예제
앞서 배운 urllib과 BeautifulSoup를 조합하면, 웹스크레이핑 및 API 요청 작업을 쉽게 수행하실 수 있습니다.

1. URL을 이용하여 웹으로부터 html을 읽어들임 (**urllib**)
2. html 분석 및 원하는 데이터를 추출 (**BeautifulSoup**)


In [20]:
from bs4 import BeautifulSoup
from urllib import request, parse

## 4.1. 네이버 금융 - 환율 정보
- 다양한 금융 정보가 공개돼 있는 "네이버 금융"에서 원/달러 환율 정보를 추출해보기
- 네이버 금융의 시장 지표 페이지 https://finance.naver.com/marketindex/
- 다음은 원/달러 환율 정보를 추출하는 프로그램임

### 1) HTML 가져오기

In [21]:
# 네이버 금융의 시장 지표 url
url = "https://finance.naver.com/marketindex/"
res = request.urlopen(url)

### 2) HTML 분석하기

In [22]:
soup = BeautifulSoup(res, "html.parser")

### 3) 원하는 데이터 추출하기

In [23]:
# soup.select_one()을 사용해 CSS 선택자로 요소 하나를 추출
price = soup.select_one("div.head_info > span.value").string
# 미국의 환전 고시 환율 값 추출
print("usd/krw =", price) 

usd/krw = 1,184.00


## 4.2. 기상청 RSS
- 기상청 RSS에서 특정 내용을 추출하는 예제
- 기상청 RSS에서 XML 데이터를 추출하고 XML 내용을 출력
- 기상청의 RSS 서비스에 지역 번호를 지정하여 데이터 요청해보기 http://www.kma.go.kr/weather/forecast/mid-term-rss3.jsp
- 참고: 기상청 RSS http://www.kma.go.kr/weather/lifenindustry/service_rss.jsp

|매개변수|의미|
|-------|----|
|stnid|기상정보를 알고 싶은 지역을 지정|

- 지역번호는 다음과 같음

|지역|지역번호|지역|지역번호|
|----|--------|----|--------|
|전국	|108	|전라북도|	146|
|서울/경기도|	109	|전라남도	|156|
|강원도	|105	|경상북도	|143|
|충청북도	|131	|경상남도	|159|
|충청남도	|133	|제주특별자치도	|184|

- 파이썬으로 요청 전용 매개변수를 만들 때는 urllib.parse 모듈의 urlencode() 함수를 사용해 매개변수를 URL로 인코딩한다.


### 1) HTML 가져오기

In [24]:
# 기상청의 RSS 서비스에 지역 번호를 지정하여 데이터 요청해보기
url = "http://www.kma.go.kr/weather/forecast/mid-term-rss3.jsp"

#매개변수를 URL로 인코딩한다. (서울/경기 지역번호 109입력)
values = {
    'stnId':'109'
}

params=parse.urlencode(values)
url += "?"+params # URL에 매개변수 추가
print("url=", url)

res = request.urlopen(url)

url= http://www.kma.go.kr/weather/forecast/mid-term-rss3.jsp?stnId=109


### 2) HTML 분석하기

In [25]:
soup = BeautifulSoup(res, "html.parser")

### 3) 원하는 데이터 추출하기

In [26]:
# find() 메서드로 원하는 부분 추출하기
header = soup.find("header")

title = header.find("title").text
wf = header.find("wf").text

print(title)
print(wf)

서울,경기도 육상중기예보
○ (강수) 29일(수) 오후에는 비가 내리겠습니다.<br />○ (기온) 이번 예보기간 아침최저기온은 14~20도로 어제(21일, 19~21도)와 비슷하거나 낮겠고, 낮최고기온은 24~27도로 어제(21일, 25~27도)와 비슷하겠습니다.<br />○ (해상) 서해중부해상의 물결은 0.5~2.0m로 일겠습니다.<br />○ (주말전망) 25일(토)과 26일(일)은 구름많겠습니다. 아침최저기온은 14~19도, 낮최고기온은 25~27도가 되겠습니다.


- css selector 기반

In [27]:
# CSS 선택자로 요소 하나 추출하기
# 위에 작성한 코드 결과와 같은 것을 확인할 수 있음.
title = soup.select_one("header > title").text
wf = header.select_one("header wf").text

print(title)
print(wf)

서울,경기도 육상중기예보
○ (강수) 29일(수) 오후에는 비가 내리겠습니다.<br />○ (기온) 이번 예보기간 아침최저기온은 14~20도로 어제(21일, 19~21도)와 비슷하거나 낮겠고, 낮최고기온은 24~27도로 어제(21일, 25~27도)와 비슷하겠습니다.<br />○ (해상) 서해중부해상의 물결은 0.5~2.0m로 일겠습니다.<br />○ (주말전망) 25일(토)과 26일(일)은 구름많겠습니다. 아침최저기온은 14~19도, 낮최고기온은 25~27도가 되겠습니다.


## 4.3. 윤동주 작가의 작품 목록

- 위키문헌 (https://ko.wikisource.org/wiki) 에 공개되어 있는 윤동주의 작품목록을 가져오기
- 윤동주 위키 (https://ko.wikisource.org/wiki/%EC%A0%80%EC%9E%90:%EC%9C%A4%EB%8F%99%EC%A3%BC)
- 하늘과 바람과 시 부분을 선택한 후 오른쪽 마우스 이용해 copy selector로 카피하면 다음의 CSS 선택자가 카피됨

**#mw-content-text > div > ul:nth-child(6) > li > b > a**

- nth-child(n) 은 n 번째 요소를 의미 즉 6번째 요소를 의미, **#mw-content-text** 내부에 있는 url 태그는 모두 작품과 관련된 태그. 따라서 따로 구분할 필요는 없으며 생략해도 됨. BeautifulSoup는 nth-child 지원하지 않음
- Recall PR7 Problem1

In [28]:
# 뒤의 인코딩 부분은 "저자:윤동주"라는 의미입니다.
# 따로 입력하지 말고 위키 문헌 홈페이지에 들어간 뒤에 주소를 복사해서 사용하세요.

# 윤동주 위키 문헌 주소
url = "https://ko.wikisource.org/wiki/%EC%A0%80%EC%9E%90:%EC%9C%A4%EB%8F%99%EC%A3%BC"
res = request.urlopen(url)
soup = BeautifulSoup(res, "html.parser")

# #mw-content-text 바로 아래에 있는 
# ul 태그 바로 아래에 있는
# li 태그 아래에 있는
# a 태그를 모두 선택합니다.(윤동주의 작품 목록 가져오기)
a_list = soup.select("#mw-content-text   ul > li  a")
for a in a_list:
    name = a.string
    print(f"- {name}", )

- 하늘과 바람과 별과 시
- 증보판
- 서시
- 자화상
- 소년
- 눈 오는 지도
- 돌아와 보는 밤
- 병원
- 새로운 길
- 간판 없는 거리
- 태초의 아침
- 또 태초의 아침
- 새벽이 올 때까지
- 무서운 시간
- 십자가
- 바람이 불어
- 슬픈 족속
- 눈감고 간다
- 또 다른 고향
- 길
- 별 헤는 밤
- 흰 그림자
- 사랑스런 추억
- 흐르는 거리
- 쉽게 씌어진 시
- 봄
- 참회록
- 간(肝)
- 위로
- 팔복
- 못자는밤
- 달같이
- 고추밭
- 아우의 인상화
- 사랑의 전당
- 이적
- 비오는 밤
- 산골물
- 유언
- 창
- 바다
- 비로봉
- 산협의 오후
- 명상
- 소낙비
- 한난계
- 풍경
- 달밤
- 장
- 밤
- 황혼이 바다가 되어
- 아침
- 빨래
- 꿈은 깨어지고
- 산림
- 이런날
- 산상
- 양지쪽
- 닭
- 가슴 1
- 가슴 2
- 비둘기
- 황혼
- 남쪽 하늘
- 창공
- 거리에서
- 삶과 죽음
- 초한대
- 산울림
- 해바라기 얼굴
- 귀뚜라미와 나와
- 애기의 새벽
- 햇빛·바람
- 반디불
- 둘 다
- 거짓부리
- 눈
- 참새
- 버선본
- 편지
- 봄
- 무얼 먹구 사나
- 굴뚝
- 햇비
- 빗자루
- 기왓장 내외
- 오줌싸개 지도
- 병아리
- 조개껍질
- 겨울
- 트루게네프의 언덕
- 달을 쏘다
- 별똥 떨어진 데
- 화원에 꽃이 핀다
- 종시


# 일반문제

In [29]:
from bs4 import BeautifulSoup
from urllib import request

# 1. 네이버 뉴스 헤드라인
배운 내용을 바탕으로 네이버 뉴스(https://news.naver.com/)에서 헤드라인 뉴스의 제목을 추출해보고자 합니다.

Q: 다음의 코드에 css selector를 추가하여 최신 기사의 헤드라인을 스크레이핑하는 코드를 완성하시오.

In [36]:
# 네이버 뉴스 주소
url = "https://news.naver.com/"

res = request.urlopen(url)
soup = BeautifulSoup(res, "html.parser")

# 최신 기사의 헤드라인을 스크레이핑
selector = "#today_main_news > div.hdline_news > ul > li > div.hdline_article_tit > a"
for a in soup.select(selector):
    title = a.text
    print(title)


                                        윤석열 측, 문준용 ‘혈세 지원’ 논평 하루 만에 철회
                                    

                                        한전, 전기료 8년 만에 올리나…내일 4분기 요금 발표
                                    

                                        이재명 "수박 기득권" 발칵…"5·18 비하하는 일베 용어"
                                    

                                        삼성·LG·현대차 美 최고 디자인상 싹쓸이
                                    

                                        윤석열 "군필자, 부동산 청약시 가점 5점 주겠다"
                                    


# 2. 시민의 소리 게시판
다음은 서울시 대공원의 시민의 소리 게시판 입니다.

https://www.sisul.or.kr/open_content/childrenpark/qna/qnaMsgList.do?pgno=1

해당 페이지에 나타난 게시글들의 제목을 수집하고자 합니다.

Q: 다음의 코드에 css selector를 추가하여 해당 페이지에서 게시글의 제목을 스크레이핑하는 코드를 완성하시오. 또한 과제 제출시 하단의 추가 내용을 참고하여 수집한 데이터를 csv 형태로 저장하여 해당 csv 파일도 함께 제출하시오.

In [31]:
# 서울시설공단 사이트 주소
url_head = "https://www.sisul.or.kr"

# 서울시설공단 시민의 소리 게시판 주소 생성
url_board = url_head + "/open_content/childrenpark/qna/qnaMsgList.do?pgno=1"


res = request.urlopen(url_board)
soup = BeautifulSoup(res, "html.parser")

# 시민의 소리 게시글의 제목을 스크레이핑 하기
# selector = "#detail_con > div.generalboard > table > tbody > tr > td.left.title > a"
selector = "#detail_con > div.generalboard > table > tbody > tr > td.left.title > a"
# 빈 리스트 생성
titles = []
links = []

# 시민의 소리 게시글의 제목과 해당 게시글 주소 리스트화하여 추출하기
for a in soup.select(selector):
    titles.append(a.text)
    links.append(url_head + a.attrs["href"])
    
print(titles, links)

['강창수 해설사님 ', '동물해설사님 칭찬', '강창수 동물 해설사님', '놀이동산 푸드코트 김치가 중국산인 이유는?', '주슨트 설명 최고예요!!', '강창수 주슨트님 최고 !!', 'ZOOCENT 스케줄표?', '호주동물 호주설명 ', '호주및 호주동물 설명에 대해 ', '동물원'] ['https://www.sisul.or.kr/open_content/childrenpark/qna/qnaMsgDetail.do;jsessionid=h7qxKDgmAP5Fe6VN7G4ytvE1zLMmxHafNke11DzIpD4ZGPyEpcX8afFjB10dTWD4.etisw2_servlet_user?qnaid=QNAS20210920000001&pgno=1', 'https://www.sisul.or.kr/open_content/childrenpark/qna/qnaMsgDetail.do;jsessionid=h7qxKDgmAP5Fe6VN7G4ytvE1zLMmxHafNke11DzIpD4ZGPyEpcX8afFjB10dTWD4.etisw2_servlet_user?qnaid=QNAS20210919000004&pgno=1', 'https://www.sisul.or.kr/open_content/childrenpark/qna/qnaMsgDetail.do;jsessionid=h7qxKDgmAP5Fe6VN7G4ytvE1zLMmxHafNke11DzIpD4ZGPyEpcX8afFjB10dTWD4.etisw2_servlet_user?qnaid=QNAS20210919000003&pgno=1', 'https://www.sisul.or.kr/open_content/childrenpark/qna/qnaMsgDetail.do;jsessionid=h7qxKDgmAP5Fe6VN7G4ytvE1zLMmxHafNke11DzIpD4ZGPyEpcX8afFjB10dTWD4.etisw2_servlet_user?qnaid=QNAS20210918000002&pgno=1', 'https://www.sisul.or.kr/open_con

## 추가 내용
수집된 자료를 데이터프레임으로 만들어 csv로 저장하는 것이 일반적입니다.

In [32]:
import pandas as pd

# 리스트화 한 titles와 links를 데이터프레임으로 생성하여 저장하기
board_df = pd.DataFrame({"title": titles, "link": links})
board_df.head()

title                                               link
0                강창수 해설사님   https://www.sisul.or.kr/open_content/childrenp...
1                동물해설사님 칭찬  https://www.sisul.or.kr/open_content/childrenp...
2              강창수 동물 해설사님  https://www.sisul.or.kr/open_content/childrenp...
3  놀이동산 푸드코트 김치가 중국산인 이유는?  https://www.sisul.or.kr/open_content/childrenp...
4            주슨트 설명 최고예요!!  https://www.sisul.or.kr/open_content/childrenp...

In [34]:
# csv로 저장하기
#index=False옵션은 위의 출력된 데이터프레임에 자동으로 추가된 인덱스 없이 저장함.
#header옵션의 별도 지정이 없으면 첫 행의 데이터가 열 이름이 됨.
board_df.to_csv("board.csv", index=False) 